### PR TITLE
Add Daffodil Developer Guide

### PR DESCRIPTION
**Pull request by tuxji**
DEVELOP.md: Provide software engineers a guide on how to develop Daffodil.  Cover the following areas of note:

  * Code Style - list coding guidelines
  * Code Workflow - list steps to contribute code to Daffodil
  * Development Environment - list tools needed to build Daffodil
  * Directory Organization - document file locations
  * Documentation - list Daffodil documentation
  * Tests - list how tests are run

DAFFODIL-2749

### CODE REVIEW

**Review by mbeckerle — APPROVED** [review]
+1 Great addition.

File: DEVELOP.md
Comment (stevedlawrence) [inline DEVELOP.md:146]: I might suggest just replacing this section with a link to the contribution workflow page? That provides steps git commands to set up a repo, which I think is easier to follow than this written prose.

I also am not sure I totally agree with the reading vs writing copy. For new contributors, there's not really a difference because they don't have write access to apache/daffodil. For them, I'd just recommend a single repo with two remotes (which is what the instructions recommend).

We may even want to create a CONTRIBUTING.md file that moves the "Contributor Workflow" from he wiki to github so it's a bit more visible.

File: DEVELOP.md
Comment (stevedlawrence) [inline DEVELOP.md:157]: We set the codecov target to `auto`, which I think just means a "patch coverage" must be the same or more than the base of the PR. Daffodil is just currently at about 80% coverage, so that's what PR's require. I"m not sure if that distinction is worth mentioning though. ~80% seems reasonable, especially since it will require a decent amount of work to make that number grow.

File: DEVELOP.md
Comment (stevedlawrence) [inline DEVELOP.md:198]: These "You must" sentences might be easier to quickly parse (and more visible) if they were a list?

I think this also more belongs in a CONTRIBUTING.md file, but it's fine here for now. Might be a good future change though.

File: DEVELOP.md
Comment (stevedlawrence) [inline DEVELOP.md:206]: Usually the person who merges the PR (i.e. an ASF comitter) marks it as resolved. I'm not even sure if non-comitters can resolve commits?

File: DEVELOP.md
Comment (stevedlawrence) [inline DEVELOP.md:215]: Maybe `s/test/verify/`? I could see someone misreading this as you add the code and someone else will add the tests, but it's important that tests be adding in the same commit that adds the code.

File: DEVELOP.md
Comment (stevedlawrence) [inline DEVELOP.md:230]: Suggest replacing 20 with "numerous" or something? As we add new directories we won't want to update this number.

File: DEVELOP.md
Comment (stevedlawrence) [inline DEVELOP.md:266]: Thoughts on removing all these `src/` lines and moving the description to the parent directory? Might make this more compact?

File: DEVELOP.md
Comment (stevedlawrence) [inline DEVELOP.md:302]: I would avoid using specific numbers, they are going to quickly go out of date and I'm not sure really provides additional value.

File: DEVELOP.md
Comment (stevedlawrence) [inline DEVELOP.md:322]: Same here, not sure these numbers are needed. It also depens on which jars you need. For example, some projects may not require the tdml-processor if they don't care about testing, or use some other infrastructure for testing.

File: DEVELOP.md
Comment (stevedlawrence) [inline DEVELOP.md:467]: While this is a great improvement, I'm not sure this sentence really belongs in this document. We don't call out how long it takes to run other tests.

File: DEVELOP.md
Comment (stevedlawrence) [inline DEVELOP.md:452]: Is it worth mentioning how to run these tests with sbt, especially since integration tests aren't run by default?

**Review by stevedlawrence — APPROVED** [review]
Big +1, this is great!

Just a number of minor comments.

File: DEVELOP.md
Comment (scholarsmate) [inline DEVELOP.md:66]: I would replace "cross-domain system devices" with "Cross-Domain Solution (CDS) devices".

File: DEVELOP.md
Comment (scholarsmate) [inline DEVELOP.md:78]: Apache Software Foundation is introduced here and later the acronym "ASF" is used.  Take the opportunity to associate Apache Software Foundation with its "ASF" acronym  here.

File: DEVELOP.md
Comment (scholarsmate) [inline DEVELOP.md:154]: In the ticket cited, once this PR is merged, we should add a reminder to update this guide once ticket has been resolved.

**Review by scholarsmate — APPROVED** [review]
+1

We need this, thank you!

File: DEVELOP.md
Comment (tuxji) [reply DEVELOP.md:419]: Yes, I've added commands showing how to run the tests.

File: DEVELOP.md
Comment (tuxji) [reply DEVELOP.md]: OK, I've removed the sentence but added how to run the integration tests.

**Pull request merged**
